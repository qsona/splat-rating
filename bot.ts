// Require the necessary discord.js classes
import { Client, GatewayIntentBits, CommandInteraction, InteractionType, Interaction, ChatInputCommandInteraction } from 'discord.js'

import registerHandler from './src/commands/register'
import newgameHandler from './src/commands/newgame'
import joinHandler from './src/commands/join'
import leaveHandler from './src/commands/leave'
import breakHandler from './src/commands/break'
import matchHandler from './src/commands/match'
import resultHandler from './src/commands/report'
import makeRegisterHandler from './src/commands/make-register'
import makeJoinHandler from './src/commands/make-join'
import makeLeaveHandler from './src/commands/make-leave'
import displayHandler from './src/commands/display'
import pingHandler from './src/commands/ping'
import separateHandler from './src/commands/separate'
import separationResetHandler from './src/commands/separation-reset'

import { tksPartyHandler, tksRecruitHandler } from './src/commands/tks'

import { execute as executeButtonHandlers } from './src/commands/buttonHandlers'
import { execute as executeModalHandlers } from './src/commands/modalHandlers'

require('dotenv').config()

const token = process.env.DISCORD_TOKEN

// Create a new client instance
const client = new Client({ intents: [GatewayIntentBits.Guilds] })

// When the client is ready, run this code (only once)
client.once('ready', () => {
  console.log('Ready!')
})

// Login to Discord with your client's token
if (token) {
  client.login(token)
} else {
  console.log('DISCORD_TOKEN is not set')
}

export interface CommandHandler {
  commandName: string
  execute: (interaction: ChatInputCommandInteraction) => Promise<void>
}

const handlers = new Map<string, CommandHandler>()

const helpHandler: CommandHandler = {
  commandName: 'sr-help',
  execute: async (interaction) => {
    await interaction.reply('参加する人は /sr-register でレーティング登録をして /sr-join で参加してね! \n参考URL: https://qsona.github.io/splat-rating/')
  },
}

;[
  registerHandler,
  newgameHandler,
  joinHandler,
  leaveHandler,
  matchHandler,
  breakHandler,
  resultHandler,
  makeRegisterHandler,
  makeJoinHandler,
  makeLeaveHandler,
  displayHandler,
  pingHandler,
  helpHandler,
  separateHandler,
  separationResetHandler,
  tksRecruitHandler,
  tksPartyHandler,
].forEach((handler) => {
  handlers.set(handler.commandName, handler)
})

client.on('interactionCreate', async (interaction) => {
  // if (interaction.type !== InteractionType.ApplicationCommand) {
  if (interaction.isChatInputCommand()) {
    const { commandName } = interaction
    console.log(commandName)
    console.log(interaction.user)

    const handler = handlers.get(commandName)
    if (handler) {
      await handler.execute(interaction)
    }
    return
  }

  if (interaction.isButton()) {
    await executeButtonHandlers(interaction)
    return
  }

  if (interaction.type === InteractionType.ModalSubmit) {
    await executeModalHandlers(interaction)
    return
  }
})
