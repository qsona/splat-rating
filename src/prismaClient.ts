import { PrismaClient } from '@prisma/client'

export const prisma = new PrismaClient({ log: ['info', 'query', 'warn', 'error'] })
