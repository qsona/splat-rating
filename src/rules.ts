export type SplatRuleSet = 'SplatZones' | 'TowerControl' | 'Rainmaker' | 'ClamBlitz' | 'TurfWar'

export const SPLAT_RULES_NAME_MAP: { code: SplatRuleSet; name: string }[] = [
  { code: 'SplatZones', name: 'ガチエリア' },
  { code: 'TowerControl', name: 'ガチヤグラ' },
  { code: 'Rainmaker', name: 'ガチホコバトル' },
  { code: 'ClamBlitz', name: 'ガチアサリ' },
  { code: 'TurfWar', name: 'ナワバリバトル' },
]

export const getRuleName = (rule: SplatRuleSet) => {
  return SPLAT_RULES_NAME_MAP.find((r) => r.code === rule)!.name
}
