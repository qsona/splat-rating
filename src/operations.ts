import assert from 'assert'
import { sum, difference } from 'lodash'
import Aigle from 'aigle'
import { JoinableGame, Rating, JoinedUser, GameMatching, PlayingGame } from '@prisma/client'
import { prisma } from './prismaClient'
import { rate } from 'openskill'
import { SplatRuleSet } from './rules'

const BETA = 200
const INITIAL_SIGMA = 200

export const createNewGame = async (userId: string, discordChannelId: string, rule: SplatRuleSet, totalGameCount: number) => {
  const rating = await prisma.rating.findUnique({
    where: {
      userId_rule: { userId, rule },
    },
  })
  if (!rating) {
    return 'RATING_DOES_NOT_EXIST'
  }

  return await prisma.$transaction(async (prisma) => {
    let joinableGame: JoinableGame
    try {
      joinableGame = await prisma.joinableGame.create({
        data: {
          discordChannelId,
          creatorUserId: userId,
          rule,
          totalGameCount,
        },
      })
    } catch (e) {
      if ((e as any).code === 'P2002') {
        return 'JOINABLE_GAME_ALREADY_EXISTS'
      }
      throw e
    }
    const joinedUser = await prisma.joinedUser.create({
      data: {
        joinableGameId: joinableGame.id,
        userId,
      },
    })

    return { rating, joinableGame, joinedUser }
  })
}

export const joinGame = async (userId: string, discordChannelId: string) => {
  // TODO: acquire lock
  const joinableGame = await prisma.joinableGame.findUnique({
    where: { discordChannelId },
  })
  if (!joinableGame) {
    return 'JOINABLE_GAME_DOES_NOT_EXIST'
  }
  const { rule } = joinableGame
  const rating = await prisma.rating.findUnique({
    where: {
      userId_rule: { userId, rule },
    },
  })
  if (!rating) {
    return 'RATING_DOES_NOT_EXIST'
  }

  return await prisma.$transaction(async (prisma) => {
    let joinedUser: JoinedUser
    try {
      joinedUser = await prisma.joinedUser.create({
        data: {
          joinableGameId: joinableGame.id,
          userId,
        },
      })
    } catch (e) {
      if ((e as any).code === 'P2002') {
        return 'USER_ALREADY_JOINED'
      }
      throw e
    }

    const joinedUsers = await prisma.joinedUser.findMany({
      where: {
        joinableGameId: joinableGame.id,
      },
    })

    if (joinedUsers.length < 8) {
      return {
        status: 'joinable' as 'joinable',
        joinedUser,
        joinableGame,
        joinedUsersCount: joinedUsers.length,
        remainUsersCount: 8 - joinedUsers.length,
        rating,
      }
    }

    // matching
    assert.equal(joinedUsers.length, 8)

    const ratings = await prisma.rating.findMany({
      where: {
        userId: { in: joinedUsers.map((u) => u.userId) },
        rule,
      },
    })
    if (ratings.length !== joinedUsers.length) {
      throw new Error(`mismatch between ratings and joinedUsers`)
    }

    const { creatorUserId } = joinableGame

    // move creator rating to first
    const creatorRatingIndex = ratings.findIndex((r) => r.userId === creatorUserId)
    if (creatorRatingIndex < 0) {
      throw new Error(`rating of creator does not exist`)
    }
    const creatorRating = ratings[creatorRatingIndex]
    ratings.splice(creatorRatingIndex, 1)
    ratings.unshift(creatorRating)

    const playingGame = await prisma.playingGame.create({
      data: {
        id: joinableGame.id,
        discordChannelId,
        rule,
        creatorUserId,
        totalGameCount: joinableGame.totalGameCount,
        currentGameCount: 0,
      },
    })

    const gameMatching = await createGameMatching(playingGame.id, ratings)

    await prisma.joinedUser.deleteMany({ where: { joinableGameId: joinableGame.id } })
    await prisma.joinableGame.delete({ where: { id: joinableGame.id } })

    return {
      status: 'matched' as 'matched',
      playingGame,
      rating,
      gameMatching,
    }
  })
}

export const createRating = async (userId: string, rule: SplatRuleSet, estimatedGachiPower: number) => {
  const rating = await prisma.rating.create({
    data: {
      mu: estimatedGachiPower,
      sigma: INITIAL_SIGMA,
      userId: userId,
      rule,
    },
  })

  return rating
}

export const createGameMatching = async (playingGameId: string, ratings: Rating[]) => {
  assert.equal(ratings.length, 8)
  const teamsRatings = calculateMatchingWithMinRateDiff(ratings)

  const gameMatching = await prisma.gameMatching.create({
    data: {
      playingGameId,
      teamsRatingIds: teamsRatings.map((teamRatings) => teamRatings.map((r) => r.id)),
    },
  })
  return gameMatching
}

export const calculateMatchingWithMinRateDiff = (ratings: Rating[]) => {
  let minRateDiff = Number.POSITIVE_INFINITY
  let rates = ratings.map((r) => r.mu)
  let totalRate = sum(rates)
  let currentAlphaTeamIndexes: number[] = []

  for (let i = 1; i <= 5; i++) {
    for (let j = i + 1; j <= 6; j++) {
      for (let k = j + 1; k <= 7; k++) {
        const alphaTeamIndexes = [0, i, j, k]
        const alphaTeamTotalRate = sum(alphaTeamIndexes.map((i) => rates[i]))
        const bravoTeamTotalRate = totalRate - alphaTeamTotalRate
        const rateDiff = Math.abs(alphaTeamTotalRate - bravoTeamTotalRate)

        if (rateDiff < minRateDiff) {
          currentAlphaTeamIndexes = alphaTeamIndexes
          minRateDiff = rateDiff
        }
      }
    }
  }

  const decidedAlphaTeamIndexes = currentAlphaTeamIndexes
  const decidedBravoTeamIndexes = difference([0, 1, 2, 3, 4, 5, 6, 7], decidedAlphaTeamIndexes)
  return [decidedAlphaTeamIndexes, decidedBravoTeamIndexes].map((teamIndexes) => {
    return teamIndexes.map((index) => ratings[index])
  })
}

export type RatingResult = {
  ratingId: string
  before: { mu: number; sigma: number }
  after: { mu: number; sigma: number }
}

export const createGameResult = async (playingGame: PlayingGame, isAlphaWin: boolean) => {
  const gameMatching = await prisma.gameMatching.findUnique({ where: { playingGameId: playingGame.id }, rejectOnNotFound: true })

  return await prisma.$transaction(async (prisma) => {
    const teamsRatings = await Aigle.map(gameMatching.teamsRatingIds as string[][], async (teamRatingIds) => {
      const teamRatings = await prisma.rating.findMany({ where: { id: { in: teamRatingIds } } })
      return teamRatingIds.map((teamRatingId) => {
        const rating = teamRatings.find((r) => r.id === teamRatingId)
        if (!rating) {
          throw new Error(`rating not found. id: ${teamRatingId}`)
        }
        return rating
      })
    })
    const sortedTeamsRatings = isAlphaWin ? [teamsRatings[0], teamsRatings[1]] : [teamsRatings[1], teamsRatings[0]]
    const newTeamsRatings = rate(sortedTeamsRatings, { beta: BETA })
    console.log('win', newTeamsRatings[0])
    console.log('lose', newTeamsRatings[1])
    const [winnerTeamsRatings, loserTeamsRatings]: RatingResult[][] = await Aigle.map(sortedTeamsRatings, async (teamRatings, teamIndex) => {
      return await Aigle.map(teamRatings, async (rating, ratingIndex) => {
        const newRating = newTeamsRatings[teamIndex][ratingIndex]
        await prisma.rating.update({
          where: {
            id: rating.id,
          },
          data: {
            mu: newRating.mu,
            sigma: newRating.sigma,
          },
        })

        return {
          ratingId: rating.id,
          before: {
            mu: rating.mu,
            sigma: rating.sigma,
          },
          after: {
            mu: newRating.mu,
            sigma: newRating.sigma,
          },
        }
      })
    })

    const gameResult = await prisma.gameResult.create({
      data: {
        beta: BETA,
        winnerTeamsRatings,
        loserTeamsRatings,
      },
    })

    await prisma.gameMatching.delete({
      where: { id: gameMatching.id },
    })

    const isFinishing = playingGame.currentGameCount + 1 === playingGame.totalGameCount

    if (isFinishing) {
      await prisma.playingGame.delete({
        where: { id: playingGame.id },
      })
      return {
        isFinished: true as true,
        gameResult,
      }
    }

    const previousRatingIds = gameMatching.teamsRatingIds as string[][]

    // create next match
    // TODO
    // const nextGameMatching = await createGameMatching(playingGame.id, [])

    return {
      isFinished: false as false,
      gameResult,
      // nextGameMatching,
    }
  })
}
