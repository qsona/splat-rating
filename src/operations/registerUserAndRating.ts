import { SplatRuleSet } from '../rules'
import { prisma } from '../prismaClient'
import { Rating, User } from '@prisma/client'

export const INITIAL_SIGMA = 200

export const registerUserAndRating = async (userId: string, username: string, guildId: string, rule: SplatRuleSet, estimatedGachiPower: number) => {
  let isNewUser = true
  let user: User | null = await prisma.user.findUnique({
    where: {
      id: userId,
    },
  })
  if (user) {
    isNewUser = false
  } else {
    try {
      user = await prisma.user.create({
        data: {
          id: userId,
          name: username,
        },
      })
    } catch (e) {
      if ((e as any).code === 'P2002') {
        // just user is already registered
        isNewUser = false
      } else {
        throw e
      }
    }
  }

  let rating: Rating | null = await prisma.rating.findUnique({
    where: {
      userId_guildId_rule: {
        userId,
        guildId,
        rule,
      },
    },
  })
  if (rating) {
    return 'RATING_ALREADY_REGISTERED' as const
  }

  try {
    rating = await prisma.rating.create({
      data: {
        guildId,
        mu: estimatedGachiPower,
        sigma: INITIAL_SIGMA,
        user: { connect: { id: userId } },
        rule,
      },
    })
  } catch (e) {
    if ((e as any).code === 'P2002') {
      return 'RATING_ALREADY_REGISTERED' as const
    }
    throw e
  }

  if (isNewUser) {
    return {
      isNewUser: true as true,
      user: user!,
      rating,
    }
  } else {
    return {
      isNewUser: false as false,
      rating,
    }
  }
}
