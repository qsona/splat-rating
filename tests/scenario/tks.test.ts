import { it } from '@jest/globals'
import { onInteractionCreated } from '../../bot'
import {
  getCustomId,
  getCustomIdWithoutData,
  createButtonInteraction,
  createButtonWithDataInteraction,
  createChatInputInteraction,
  getModalCustomId,
  createModalWithDataInteraction,
  createModalInteraction,
  getModalCustomIdWithoutData,
} from './helper'
import { prisma } from '../../src/prismaClient'
import { registerUserAndRating } from '../../src/operations/registerUserAndRating'
import { tksCreateParty } from '../../src/operations/tksCreateParty'
import assert from 'assert'
import { SplatRuleSet } from '../../src/rules'

const guildId = 'guild1'
const channelId = 'channel1'

describe('tks scenario', () => {
  it('tks scenario', async () => {
    const rule = 'SplatZones'
    // prepare
    await registerUserAndRating('user1', 'username1', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user2', 'username2', guildId, 'SplatZones', 2100)
    await registerUserAndRating('user3', 'username3', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user4', 'username4', guildId, 'SplatZones', 2300)
    await registerUserAndRating('user5', 'username5', guildId, 'SplatZones', 2400)
    await registerUserAndRating('user6', 'username6', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user7', 'username7', guildId, 'SplatZones', 2100)
    await registerUserAndRating('user8', 'username8', guildId, 'SplatZones', 1900)

    // new party
    await onInteractionCreated(
      createChatInputInteraction(
        'tks-party',
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
          options: {
            getMentionable: (arg: string) =>
              arg === 'user2'
                ? { id: 'user2', username: 'username2' }
                : arg === 'user3'
                ? { id: 'user3', username: 'username3' }
                : arg === 'user4'
                ? { id: 'user4', username: 'username4' }
                : null,
          },
        },
        {
          onReply: async (options) => {
            expect(options).toMatchSnapshot()
          },
          onChannelSend: async (options) => {
            expect(options.components).toHaveLength(3)
            const party = await prisma.tksParty.findFirst()
            expect(getCustomId(options.components![0])).toBe(`button-tks-find-opponent@${party!.id}`)
            expect(getCustomId(options.components![1])).toBe(`button-tks-set-team-name@${party!.teamId}`)
            expect(getCustomId(options.components![2])).toBe(`button-tks-break-party@${party!.teamId}`)
          },
        }
      )
    )

    const firstParty = await prisma.tksParty.findFirst({ include: { team: { include: { tksTeamUsers: true } } } })
    expect(firstParty).toBeTruthy()
    expect(firstParty?.team).toBeTruthy()
    expect(firstParty?.team.tksTeamUsers).toHaveLength(4)

    const firstPartyRatingInitial = await prisma.tksRating.findFirst({ where: { teamId: firstParty!.teamId, rule } })
    expect(firstPartyRatingInitial).toBeTruthy()

    // new recruit
    await onInteractionCreated(
      createChatInputInteraction(
        'tks-recruit',
        {
          channelId,
          guildId,
          user: {
            id: 'user5',
            username: 'username5',
          },
        },
        {
          onShowModal: async (options) => {
            expect(getModalCustomId(options)).toBe('modal-tks-recruit')
          },
        }
      )
    )

    await onInteractionCreated(
      createModalInteraction(
        'modal-tks-recruit',
        {
          channelId,
          guildId,
          user: {
            id: 'user5',
            username: 'username5',
          },
          fields: {
            getTextInputValue: () => 'description 1', // TODO: use mock
          },
        },
        {
          onReply: async (options) => {
            expect(options.content).toMatchSnapshot()
            expect(options.components).toHaveLength(2)
            const room = await prisma.tksRecruitingRoom.findFirst()
            expect(getCustomId(options.components![0])).toBe(`button-tks-room-join@${room!.id}`)
            expect(getCustomId(options.components![1])).toBe(`button-tks-break-room@${room!.id}`)
          },
        }
      )
    )
    const recruitingRoom = await prisma.tksRecruitingRoom.findFirst({ include: { recruitingRoomUsers: true } })
    expect(recruitingRoom).toBeTruthy()
    expect(recruitingRoom!.description).toBe('description 1')
    expect(recruitingRoom!.recruitingRoomUsers).toHaveLength(1)
    expect(recruitingRoom!.recruitingRoomUsers[0].userId).toBe('user5')

    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-room-join',
        recruitingRoom!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user6',
            username: 'username6',
          },
        },
        {
          onReply: async (options) => {
            expect(options.components).toHaveLength(2)
            expect(getCustomId(options.components![0])).toBe(`button-tks-room-join@${recruitingRoom!.id}`)
            expect(getCustomId(options.components![1])).toBe(`button-tks-leave-room@${recruitingRoom!.id}`)
          },
        }
      )
    )

    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-room-join',
        recruitingRoom!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user7',
            username: 'username7',
          },
        },
        {
          onReply: async (options) => {
            expect(options.components).toHaveLength(2)
            expect(getCustomId(options.components![0])).toBe(`button-tks-room-join@${recruitingRoom!.id}`)
            expect(getCustomId(options.components![1])).toBe(`button-tks-leave-room@${recruitingRoom!.id}`)
          },
        }
      )
    )

    // 4人目, party 結成
    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-room-join',
        recruitingRoom!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user8',
            username: 'username8',
          },
        },
        {
          onReply: async (options) => {
            expect(options).toMatchSnapshot()
          },
          onChannelSend: async (options) => {
            expect(options.content).toMatchSnapshot()
            expect(options.components).toHaveLength(3)
            expect(getCustomIdWithoutData(options.components![0])).toBe(`button-tks-find-opponent`)
            expect(getCustomIdWithoutData(options.components![1])).toBe(`button-tks-set-team-name`)
            expect(getCustomIdWithoutData(options.components![2])).toBe(`button-tks-break-party`)
          },
        }
      )
    )

    // party1 が対戦相手を探す
    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-find-opponent',
        firstParty!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
        },
        {
          onShowModal: async (options) => {
            expect(getModalCustomIdWithoutData(options)).toBe('modal-tks-find-opponent')
          },
        }
      )
    )
    await onInteractionCreated(
      createModalWithDataInteraction(
        'modal-tks-find-opponent',
        firstParty!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
          fields: { getTextInputValue: () => '3' }, // TODO: use mock
        },
        {
          onReply: async (options) => {
            expect(options.content).toMatchSnapshot()
            expect(options.components).toHaveLength(1)
            expect(getCustomId(options.components![0])).toBe(`button-tks-match@${firstParty!.id}`)
          },
        }
      )
    )
    const findingOpponent = await prisma.tksFindingOpponent.findFirst()
    expect(findingOpponent).toBeTruthy()
    expect(findingOpponent!.winCountOfMatch).toBe(3)

    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-match',
        firstParty!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user5',
            username: 'username5',
          },
        },
        {
          onReply: async (options) => {
            expect(options.content).toMatchSnapshot()
            expect(options.components).toHaveLength(1)
            expect(getCustomIdWithoutData(options.components![0])).toBe(`button-tks-report`)
          },
        }
      )
    )
    expect(await prisma.tksFindingOpponent.findFirst()).toBeNull()
    const match = await prisma.tksMatch.findFirst()
    expect(match).toBeTruthy()
    expect(match!.primaryTeamId).toBe(firstParty!.teamId)

    // report
    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-report',
        match!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user2',
            username: 'username2',
          },
        },
        {
          onShowModal: async (options) => {
            expect(getModalCustomId(options)).toBe(`modal-tks-report@${match!.id}`)
          },
        }
      )
    )
    await onInteractionCreated(
      createModalWithDataInteraction(
        'modal-tks-report',
        match!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user2',
            username: 'username2',
          },
          fields: { getTextInputValue: (arg: string) => (arg === 'primaryWinCount' ? '3' : '2') }, // TODO: use mock
        },
        {
          onReply: async (options) => {
            expect(options.content).toMatchSnapshot()
            expect(options.components || []).toHaveLength(0)
          },
        }
      )
    )
    expect(await prisma.tksMatch.findFirst()).toBeNull()
    const matchResult = await prisma.tksMatchResult.findFirst()
    expect(matchResult).toBeTruthy()
    expect(matchResult!.primaryTeamId).toBe(firstParty!.teamId)
    expect(matchResult!.primaryWinCount).toBe(3)
    expect(matchResult!.opponentWinCount).toBe(2)

    const firstPartyRating = await prisma.tksRating.findFirst({
      where: { teamId: firstParty!.teamId, rule: 'SplatZones' },
    })
    expect(firstPartyRating).toBeTruthy()
    expect(firstPartyRating!.mu).toBeGreaterThan(firstPartyRatingInitial!.mu)
    expect(firstPartyRating!.sigma).toBeLessThan(firstPartyRatingInitial!.sigma)
  })

  it('should work with 2 matches', async () => {
    const rule: SplatRuleSet = 'SplatZones'
    await registerUserAndRating('user1', 'username1', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user2', 'username2', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user3', 'username3', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user4', 'username4', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user5', 'username5', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user6', 'username6', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user7', 'username7', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user8', 'username8', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user9', 'username1', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user10', 'username2', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user11', 'username3', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user12', 'username4', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user13', 'username5', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user14', 'username6', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user15', 'username7', guildId, 'SplatZones', 2000)
    await registerUserAndRating('user16', 'username8', guildId, 'SplatZones', 2000)

    const { party: party1 } = await tksCreateParty(['user1', 'user2', 'user3', 'user4'], guildId)
    const { party: party2 } = await tksCreateParty(['user5', 'user6', 'user7', 'user8'], guildId)
    const { party: party3 } = await tksCreateParty(['user9', 'user10', 'user11', 'user12'], guildId)
    const { party: party4 } = await tksCreateParty(['user13', 'user14', 'user15', 'user16'], guildId)
    assert(party1)
    assert(party2)
    assert(party3)
    assert(party4)

    // party 1, party 4 が募集
    await onInteractionCreated(
      createModalWithDataInteraction(
        'modal-tks-find-opponent',
        party1!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
          fields: { getTextInputValue: () => '3' }, // TODO: use mock
        },
        {
          onReply: async (options) => {
            expect(options.components).toHaveLength(1)
            expect(getCustomId(options.components![0])).toBe(`button-tks-match@${party1!.id}`)
          },
        }
      )
    )
    await onInteractionCreated(
      createModalWithDataInteraction(
        'modal-tks-find-opponent',
        party4!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user13',
            username: 'username13',
          },
          fields: { getTextInputValue: () => '3' }, // TODO: use mock
        },
        {
          onReply: async (options) => {
            expect(options.components).toHaveLength(1)
            expect(getCustomId(options.components![0])).toBe(`button-tks-match@${party4!.id}`)
          },
        }
      )
    )

    // party3 => party4 に申しこみ
    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-match',
        party4!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user9',
            username: 'username9',
          },
        },
        {
          onReply: async (options) => {
            expect(options.components).toHaveLength(1)
            const match = await prisma.tksMatch.findFirst({ where: { primaryTeamId: party4.teamId, opponentTeamId: party3.teamId } })
            expect(match).toBeTruthy()
            expect(getCustomId(options.components![0])).toBe(`button-tks-report@${match!.id}`)
          },
        }
      )
    )

    // party 2 => party1 に申し込み
    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-match',
        party1!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user5',
            username: 'username5',
          },
        },
        {
          onReply: async (options) => {
            expect(options.components).toHaveLength(1)
            const match = await prisma.tksMatch.findFirst({ where: { primaryTeamId: party1.teamId, opponentTeamId: party2.teamId } })
            expect(match).toBeTruthy()
            expect(getCustomId(options.components![0])).toBe(`button-tks-report@${match!.id}`)
          },
        }
      )
    )

    const match43 = await prisma.tksMatch.findFirst({ where: { primaryTeamId: party4.teamId, opponentTeamId: party3.teamId } })!
    const match12 = await prisma.tksMatch.findFirst({ where: { primaryTeamId: party1.teamId, opponentTeamId: party2.teamId } })!

    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-report',
        match12!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
        },
        {
          onShowModal: async (options) => {
            expect(getModalCustomId(options)).toBe(`modal-tks-report@${match12!.id}`)
          },
        }
      )
    )
    await onInteractionCreated(
      createModalWithDataInteraction(
        'modal-tks-report',
        match12!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user1',
            username: 'username1',
          },
          fields: { getTextInputValue: (arg: string) => (arg === 'primaryWinCount' ? '3' : '2') }, // TODO: use mock
        },
        {
          onReply: async (options) => {
            expect(options.components || []).toHaveLength(0)
          },
        }
      )
    )
    await onInteractionCreated(
      createButtonWithDataInteraction(
        'button-tks-report',
        match43!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user13',
            username: 'username13',
          },
        },
        {
          onShowModal: async (options) => {
            expect(getModalCustomId(options)).toBe(`modal-tks-report@${match43!.id}`)
          },
        }
      )
    )
    await onInteractionCreated(
      createModalWithDataInteraction(
        'modal-tks-report',
        match43!.id,
        {
          channelId,
          guildId,
          user: {
            id: 'user13',
            username: 'username13',
          },
          fields: { getTextInputValue: (arg: string) => (arg === 'primaryWinCount' ? '2' : '3') }, // TODO: use mock
        },
        {
          onReply: async (options) => {
            expect(options.components || []).toHaveLength(0)
          },
        }
      )
    )
    expect(await prisma.tksMatch.findFirst()).toBeNull()
    const matchResult12 = await prisma.tksMatchResult.findFirst({ where: { primaryTeamId: party1.teamId, opponentTeamId: party2.teamId } })
    expect(matchResult12).toBeTruthy()
    expect(matchResult12!.primaryWinCount).toBe(3)
    expect(matchResult12!.opponentWinCount).toBe(2)

    const matchResult43 = await prisma.tksMatchResult.findFirst({ where: { primaryTeamId: party4.teamId, opponentTeamId: party3.teamId } })
    expect(matchResult43).toBeTruthy()
    expect(matchResult43!.primaryWinCount).toBe(2)
    expect(matchResult43!.opponentWinCount).toBe(3)

    // win: party 1, lose: party 2
    // win: party 3, lose: party 4

    const party1Rating = await prisma.tksRating.findFirst({
      where: { teamId: party1!.teamId, rule: 'SplatZones' },
    })
    expect(party1Rating).toBeTruthy()
    expect(party1Rating!.mu).toBeGreaterThan(2000)
    expect(party1Rating!.playCount).toBe(1)
    expect(party1Rating!.winCount).toBe(1)
    expect(party1Rating!.loseCount).toBe(0)

    const party2Rating = await prisma.tksRating.findFirst({
      where: { teamId: party2!.teamId, rule: 'SplatZones' },
    })
    expect(party2Rating).toBeTruthy()
    expect(party2Rating!.mu).toBeLessThan(2000)
    expect(party2Rating!.playCount).toBe(1)
    expect(party2Rating!.winCount).toBe(0)
    expect(party2Rating!.loseCount).toBe(1)

    const party3Rating = await prisma.tksRating.findFirst({
      where: { teamId: party3!.teamId, rule: 'SplatZones' },
    })
    expect(party3Rating).toBeTruthy()
    expect(party3Rating!.mu).toBeGreaterThan(2000)
    expect(party3Rating!.playCount).toBe(1)
    expect(party3Rating!.winCount).toBe(1)
    expect(party3Rating!.loseCount).toBe(0)

    const party4Rating = await prisma.tksRating.findFirst({
      where: { teamId: party4!.teamId, rule: 'SplatZones' },
    })
    expect(party4Rating).toBeTruthy()
    expect(party4Rating!.mu).toBeLessThan(2000)
    expect(party4Rating!.playCount).toBe(1)
    expect(party4Rating!.winCount).toBe(0)
    expect(party4Rating!.loseCount).toBe(1)
  })
})
