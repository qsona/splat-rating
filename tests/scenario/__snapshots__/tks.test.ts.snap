// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tks scenario tks scenario 1`] = `
{
  "content": "username1 username2 username3 username4 がパーティーを結成したぞ!
[計測中...0/5] (推定R2000) | (チーム名未設定)
メンバーは<#1043583020457807982> に移動してください 🚀",
}
`;

exports[`tks scenario tks scenario 2`] = `
"通知切ってますなう
username5: 対抗戦味方募集@3
description 1"
`;

exports[`tks scenario tks scenario 3`] = `
{
  "content": "username8 が参加しました。
対抗戦味方募集@うまり
[計測中...0/5] (推定R2000) | (チーム名未設定)
メンバーは<#1043583020457807982> に移動してください 🚀",
}
`;

exports[`tks scenario tks scenario 4`] = `
"username5 username6 username7 username8 がパーティーを結成したぞ!
[計測中...0/5] (推定R2000) | (チーム名未設定)"
`;

exports[`tks scenario tks scenario 5`] = `
"対抗戦相手募集
メンバー: username1 username2 username3 username4
ルール: ガチエリア 3本先取
3"
`;

exports[`tks scenario tks scenario 6`] = `
"対抗戦開始 🚀
アルファ: [チーム名: (未定)] username1 username2 username3 username4
ブラボー: [チーム名: (未定)] username5 username6 username7 username8

結果報告はアルファチームが行ってください 💪"
`;

exports[`tks scenario tks scenario 7`] = `
"アルファ: [チーム名: (未定)] username1 username2 username3 username4
ブラボー: [チーム名: (未定)] username5 username6 username7 username8

結果: アルファ 3 - 2 ブラボー 💡"
`;
